openfeature_flagsmith/provider.py

```diff
@@ -0,0 +1,134 @@
+        self._client = client
+        self.return_value_for_disabled_flags = return_value_for_disabled_flags
+        self.use_flagsmith_defaults = use_flagsmith_defaults
+        self.use_boolean_config_value = use_boolean_config_value
+
+    def get_metadata(self) -> Metadata:
+        return Metadata(name="FlagsmithProvider")
+
+    def resolve_boolean_details(
+        self,
+        key: str,
+        default_value: bool,
+        evaluation_context: EvaluationContext = EvaluationContext(),
+    ) -> FlagResolutionDetails[bool]:
+        return self._resolve(key, FlagType.BOOLEAN, default_value, evaluation_context)
+
+    def resolve_string_details(
+        self,
+        key: str,
+        default_value: str,
+        evaluation_context: EvaluationContext = EvaluationContext(),
+    ) -> FlagResolutionDetails[str]:
+        return self._resolve(key, FlagType.STRING, default_value, evaluation_context)
+
+    def resolve_integer_details(
+        self,
+        key: str,
+        default_value: Number,
+        evaluation_context: EvaluationContext = EvaluationContext(),
+    ) -> FlagResolutionDetails[int]:
+        return self._resolve(key, FlagType.INTEGER, default_value, evaluation_context)
+
+    def resolve_float_details(
+        self,
+        key: str,
+        default_value: Number,
+        evaluation_context: EvaluationContext = EvaluationContext(),
+    ) -> FlagResolutionDetails[float]:
+        return self._resolve(key, FlagType.FLOAT, default_value, evaluation_context)
+
+    def resolve_object_details(
+        self,
+        key: str,
+        default_value: dict,
+        evaluation_context: EvaluationContext = EvaluationContext(),
+    ) -> FlagResolutionDetails[typing.Union[dict, list]]:
+        return self._resolve(key, FlagType.OBJECT, default_value, evaluation_context)
+
+    def _resolve(
+        self,
+        key: str,
+        flag_type: FlagType,
+        default_value: typing.Any,
+        evaluation_context: EvaluationContext,
+    ) -> FlagResolutionDetails:
+        try:
+            flag = self._get_flags(evaluation_context).get_flag(key)
+        except FlagsmithClientError:
+            return FlagResolutionDetails(
+                value=default_value,
+                error_code=ErrorCode.GENERAL,
+                reason=Reason.ERROR,
+            )
+
+        if flag.is_default and not self.use_flagsmith_defaults:
+            return FlagResolutionDetails(
+                value=default_value,
+                error_code=ErrorCode.FLAG_NOT_FOUND,
+                reason=Reason.ERROR,
+            )
+
+        if flag_type == FlagType.BOOLEAN and not self.use_boolean_config_value:
+            return FlagResolutionDetails(value=flag.enabled)
+
+        if not (self.return_value_for_disabled_flags or flag.enabled):
+            return FlagResolutionDetails(
+                value=default_value,
+                error_code=ErrorCode.GENERAL,
+                reason=Reason.DISABLED,
+            )
```

Comment:
Instead of returning an resolution with errors, it's better to throw, which will result in the error hooks being run in the SDK" https://github.com/open-feature/python-sdk/blob/7ba7d6146f0f801cadfd7593dc6df4b7d4f488d4/openfeature/client.py#L337-L342

This is something we could do better in many SDKs - make it clear that you should only return successes and otherwise throw some instance of `OpenFeatureError`. We could possibly do this by making it impossible for provider authors to set error codes in `FlagResolutionDetails` - not sure.

cc @federicobond

Reply:
Thanks for this Todd. I believe this is done for all the instances where we were directly returning FlagResolutionDetails with an error_code.